### PR TITLE
Refactor HTML parser: lexer-based architecture with improved Twig support

### PR DESCRIPTION
**Pull request by shyim**
## Summary

This PR replaces the old recursive-descent HTML parser with a new lexer-based architecture that provides better separation of concerns, improved error handling, and more robust Twig template support. The old `internal/twigparser` package is removed entirely in favor of the enhanced `internal/html` package.

### Architecture Refactoring
- **New lexer layer** (`lexer.go`, `tokens.go`, `pos.go`): Single-pass tokenizer that recognizes HTML elements, attributes, and Twig delimiters (`{% %}`, `{{ }}`, `{# #}`) with precise position tracking
- **Token-driven parser** (`parser.go`): Replaces recursive descent with a cleaner state machine that walks the token stream, making it easier to handle complex nesting and error recovery
- **Separated formatting logic** (`format.go`): All Dump/rendering code moved to a dedicated file, making the parser logic independent of output formatting

### Twig Tag System
- **Tag registry** (`tags.go`): Pluggable tag parser system with `TagSpec` for declarative tag definitions
- **Dedicated tag parsers**: Each Twig tag type (block, if, for, set, verbatim, etc.) has its own file (`tag_*.go`) with focused parsing logic
- **Improved if/elseif/else handling** (`tag_if.go`): Proper branch parsing with correct terminator detection
- **Generic block support** (`tag_generic.go`): Handles unknown/future Twig tags gracefully by preserving them as raw text

### AST Improvements
- **Cleaner node types** (`ast.go`): Consolidated node definitions with clear interfaces
- **Better context awareness**: `nodeContext` enum distinguishes top-level parsing from element-children parsing, enabling smarter whitespace handling
- **Inline-mixed content detection**: Preserves formatting of embedded expressions (e.g., `{% for %}{{ x }}{% endfor %}`) by detecting when a block contains only text + expressions

### Error Handling
- **Structured errors** (`errors.go`): `ParseError` with source location, filename, and related locations for better diagnostics
- **Position tracking**: Every token carries precise line/column information for accurate error reporting

### Testing & Validation
- **Smoke tests** (`smoke_test.go`): Can parse real Shopware storefront templates when `HTML_SMOKE_CORPUS` is set
- **Fuzz testing** (`fuzz_test.go`): Lexer fuzzed with fixture inputs to catch edge cases
- **Fixture-based tests** (`lexer_test.go`): Validates lexer against testdata files

## Notable Implementation Details

- **Whitespace handling**: Top-level nodes trim whitespace before deciding whether to emit RawNodes; element children preserve all whitespace for faithful formatting
- **Inline-mixed detection**: Blocks containing only text + template expressions flow children verbatim to avoid re-indenting on every format pass
- **Void element recognition**: `isVoidElement()` correctly identifies self-closing HTML tags
- **Backward compatibility**: `NewParser()`, `NewAdminParser()`, and `NewStorefrontParser()` entry points maintain the existing API; `ConfiguredNodeList` wraps nodes with their formatting config

## Removed

- `internal/twigparser/` package entirely (twig.go, nodes.go, node_list.go, parser_types.go, twig_test.go)
- Old parser helpers and attribute encoding logic (now in format.go)

## Updated

- `internal/verifier/admin_twig.go` and `storefront_twig.go`: Now use `NewAdminParser()` and `NewStorefrontParser()` from the new html package

https://claude.ai/code/session_019e6qYC1a5j1TStxyedAaEy

### CODE REVIEW

File: internal/html/parser.go [hunk 1]
```diff
@@ -1,1642 +1,499 @@
...
-			builder.WriteString("\n")
-			for j, child := range nonEmptyChildren {
-				if elementChild, ok := child.(*ElementNode); ok {
-					builder.WriteString(elementChild.Dump(indent + 1))
-				} else {
-					for i := 0; i < indent+1; i++ {
-						builder.WriteString(indentStr)
+			node.Attributes = append(node.Attributes, attr)
+
+		case tokTwigStmtOpen:
```
Comment (shyim) [inline internal/html/parser.go:440]: This drops Twig that appears inside an opening HTML tag unless it is an `{% if %}` statement. The lexer already emits `tokTwigExprOpen` for cases like `<div {{ attributes }}>`, but the attribute parser only handles `tokTwigStmtOpen` and the default branch later just advances. Any formatter/fixer pass would remove those dynamic attributes from the template. Please preserve unknown Twig in attribute context as raw/structured attribute nodes instead of skipping it.

File: internal/html/tag_more.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// Built-in Twig tags. Tags with their own non-trivial parser (block, if,
+// parent, set, verbatim, twig comment) live in dedicated files. Everything
+// else funnels through the makeBlockTagParser / makeStandaloneTagParser
+// helpers.
+//
+// Custom project-specific tags should be registered from the caller's own
+// init() via RegisterBlockTag / RegisterStandaloneTag — see tags.go.
+func init() {
+	// Block tags (body + EndTag).
+	RegisterBlockTag("for", "endfor", "else")
```
Comment (shyim) [inline internal/html/tag_more.go:12]: Registering `for` with `else` as a follower currently makes valid Twig fail. `parseNodesUntil` returns `stopIfTerminator` for followers, but `makeBlockTagParser` only accepts `stopGenericEndTag`, so `{% for item in items %}...{% else %}...{% endfor %}` reports a missing `endfor`. Either teach generic block parsing how to consume follower branches or do not register `else` here until it is supported.

File: internal/html/parser.go [hunk 3]
```diff
@@ -1,1642 +1,499 @@
...
+	switch strings.ToLower(tag) {
+	case "area", "base", "br", "col", "embed", "hr", "img", "input", "keygen", "link", "meta", "param", "source", "track", "wbr":
+		return true
 	}
+	return false
+}
 
-	if a.Value == "" {
-		return builder.String() + a.Key
+func TraverseNode(n NodeList, f func(*ElementNode)) {
```
Comment (shyim) [inline internal/html/parser.go:24]: `TraverseNode` needs to recurse into `TwigGenericBlockNode` now that this parser creates generic blocks for `for`, `embed`, `macro`, `apply`, and `with`. The downstream Twig linters/fixers rely on `TraverseNode`, so elements inside those blocks will no longer be checked or fixed.

File: internal/html/format.go [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+	for i := 0; i < indent; i++ {
+		builder.WriteString(indentStr)
+	}
+
+	builder.WriteString("<!-- " + c.Text + " -->")
+
+	return builder.String()
+}
+
+func (t *TemplateExpressionNode) Dump(indent int) string {
```
Comment (shyim) [inline internal/html/format.go:200]: The lexer tracks Twig whitespace-control delimiters (`{%-`, `-%}`, `{{-`, `-}}`), but the AST/formatter discards that information and always emits normal delimiters. That changes Twig rendering semantics after a format/fix pass. The trim flags need to be represented on the relevant nodes and emitted by `Dump` for expressions, statements, and block boundaries.

**Review by shyim — COMMENTED** [review]
Inline comments from local branch review.
